src/qibolab/transpilers/general_connectivity.py

```diff
@@ -0,0 +1,352 @@
+        backward = path[meeting_point + 1 :]
+        if len(forward) > 1:
+            for i in range(len(forward) - 1):
+                self.transpiled_circuit.add(gates.SWAP(self.qubit_map[forward[i]], self.qubit_map[forward[i + 1]]))
+        if len(backward) > 1:
+            for i in range(len(backward) - 1, 0, -1):
+                self.transpiled_circuit.add(gates.SWAP(self.qubit_map[backward[i]], self.qubit_map[backward[i - 1]]))
+        return
+
+    def update_qubit_map(self):
+        """Update the qubit mapping after adding swaps"""
+        old_mapping = deepcopy(self.qubit_map)
+        for key in self.mapping.keys():
+            self.qubit_map[self.mapping[key]] = old_mapping[key]
+
+    def transpile(self, qibo_circuit):
+        """Qubit initialization and circuit transpiling.
+
+        Args:
+            qibo_circuit: circuit to be transpiled.
+
+        Returns:
+            hardware_mapped_circuit: circut mapped to hardware topology.
+            final_mapping (dict): final qubit mapping.
+            init_mapping (dict): initial qubit mapping.
+            added_swaps (int): number of swap gates added.
+        """
+        self.circuit_repr = self.translate_circuit(qibo_circuit)
+        keys = list(self.connectivity.nodes())
+        if self.init_method == "greedy":
+            self.graph, self.mapping = self.greedy_init()
+        elif self.init_method == "subgraph":
+            self.graph, self.mapping = self.subgraph_init()
+        elif self.init_method == "custom":
+            self.mapping = {keys[i]: self.mapping["q" + str(i)] for i in range(len(keys))}
+            self.graph = nx.relabel_nodes(self.connectivity, self.mapping)
+        else:
+            print("ERROR")
+        init_qubit_map = self.init_qubit_map(self.mapping)
+        init_mapping = {keys[i]: init_qubit_map[i] for i in range(len(keys))}
+        self.qubit_map = np.sort(init_qubit_map)
+        self.transpiled_circuit = self.init_circuit(qibo_circuit)
+        len_2q_circuit = len(self.circuit_repr)
+        self.circuit_repr = self.reduce(self.graph)
+        self.pos = 0
+        matched_2q_gates = len_2q_circuit - len(self.circuit_repr)
+        self.add_gates(qibo_circuit, matched_2q_gates)
+        self.n_swap = 0
+        while len(self.circuit_repr) != 0:
+            len_2q_circuit = len(self.circuit_repr)
+            path, meeting_point = self.relocate()
+            matched_2q_gates = len_2q_circuit - len(self.circuit_repr)
+            self.add_swaps(path, meeting_point)
+            self.update_qubit_map()
+            self.add_gates(qibo_circuit, matched_2q_gates)
+        hardware_mapped_circuit = self.init_mapping_circuit(self.transpiled_circuit, init_qubit_map)
+        final_mapping = {keys[i]: init_qubit_map[self.qubit_map[i]] for i in range(len(keys))}
+        return hardware_mapped_circuit, final_mapping, init_mapping, self.n_swap
+
+    def set_special_connectivity(self, connectivity):
```

Comment:
This is fine for initial testing but before merging this I would prefer to have a proper mechanism to set the connectivity from the platform, instead of hardcoding it here. The idea is that everything in the core code of qibolab, including instruments, pulses, transpiler, etc. should be agnostic of any lab specifics. Qibolab provides only the primitives to define and control different platforms, it does not provide any actual instance of such a platform.

For convenience (and testing) we define specific platform instances in the `create_*` functions in platform.py and the runcards. `set_special_connectivity` should be moved to one of these places. Since it is using networkx, it is simpler to do this in platform.py, instead of the runcard which is yaml. The connectivity does not change during calibration so it does not need to be in the runcard anyway.

More specifically the pipeline could be as follows:
* `connectivity` is defined as `platform` attribute in `create_*` method in platform.py. It is NOT read from the runcard.
* `backend.execute_circuit` (which has access to the `platform`) calls the `Transpiler` and passes `platform.connectivity`.
* `Transpiler` also has a user-friendly API to allow users to transpile circuits without necessarily executing them on hardware. Note that the users can also use the pre-coded TII platforms to pass their connectivities to the `Transpiler` even without having access to the actual instruments.

@scarrazza @andrea-pasquale let me know if you agree.